x-pack/legacy/plugins/infra/public/store/v2/log_entries/index.ts

```diff
@@ -0,0 +1,240 @@
+};
+
+enum ShouldFetchMoreEntries {
+  Before,
+  After,
+}
+
+const shouldFetchMoreEntries = ({ pagesAfterEnd, pagesBeforeStart }: FetchMoreEntriesParams) => {
+  if (pagesBeforeStart === null || pagesAfterEnd === null) return false;
+  if (pagesBeforeStart < DESIRED_BUFFER_PAGES) return ShouldFetchMoreEntries.Before;
+  if (pagesAfterEnd < DESIRED_BUFFER_PAGES) return ShouldFetchMoreEntries.After;
+  return false;
+};
+
+const useFetchEntriesEffect = (
+  dispatch: Dispatch,
+  state: LogEntriesState,
+  params: FetchEntriesParams
+) => {
+  const { getLogEntriesAround, getLogEntriesBefore, getLogEntriesAfter } = useGraphQLQueries();
+
+  const [prevParams, cachePrevParams] = useState(params);
+
+  const runFetchNewEntriesRequest = async () => {
+    dispatch({ type: Action.FetchingNewEntries });
+    try {
+      const payload = await getLogEntriesAround(params);
+      dispatch({ type: Action.ReceiveNewEntries, payload });
+    } catch (e) {
+      dispatch({ type: Action.ErrorOnNewEntries });
+    }
+  };
+
+  const runFetchMoreEntriesRequest = async (direction: ShouldFetchMoreEntries) => {
+    dispatch({ type: Action.FetchingMoreEntries });
+    const getEntriesBefore = direction === ShouldFetchMoreEntries.Before;
+    const getMoreLogEntries = getEntriesBefore ? getLogEntriesBefore : getLogEntriesAfter;
+    try {
+      const payload = await getMoreLogEntries(params);
+      dispatch({
+        type: getEntriesBefore ? Action.ReceiveEntriesBefore : Action.ReceiveEntriesAfter,
+        payload,
+      });
+    } catch (e) {
+      dispatch({ type: Action.ErrorOnMoreEntries });
+    }
+  };
+
+  const fetchNewEntriesEffectDependencies = Object.values(
+    pick(params, ['sourceId', 'filterQuery', 'timeKey'])
+  );
+  const fetchNewEntriesEffect = () => {
+    if (shouldFetchNewEntries({ ...params, ...state, prevParams })) {
+      runFetchNewEntriesRequest();
+    }
+    cachePrevParams(params);
+  };
+
+  const fetchMoreEntriesEffectDependencies = Object.values(
+    pick(params, ['pagesAfterEnd', 'pagesBeforeStart'])
+  );
+  const fetchMoreEntriesEffect = () => {
+    const direction = shouldFetchMoreEntries(params);
+    switch (direction) {
+      case ShouldFetchMoreEntries.Before:
+      case ShouldFetchMoreEntries.After:
+        runFetchMoreEntriesRequest(direction);
+        break;
+      default:
+        break;
+    }
+  };
+  useEffect(fetchNewEntriesEffect, fetchNewEntriesEffectDependencies);
+  useEffect(fetchMoreEntriesEffect, fetchMoreEntriesEffectDependencies);
+
+  return {
+    fetchNewerEntries: useCallback(() => runFetchMoreEntriesRequest(ShouldFetchMoreEntries.After), [
+      params,
+    ]),
+  };
+};
+
+export const useLogEntriesStore: () => [
+  LogEntriesState,
+  (deps: LogEntriesDependencies) => void,
+  LogEntriesCallbacks
+] = () => {
+  const [state, dispatch] = useReducer(logEntriesStateReducer, logEntriesInitialState);
+
+  const [dependencies, subscriber] = useState(logEntriesInitialDependencies);
```

Comment:
I'm not sure I understand how this works.

My brain expects `dependencies` to be used to invalidate a hook somewhere else, but if I understand correctly they are used as parameters for the request to fetch the logs.

Maybe the name is too generic. I think a more specific naming (although more verbose) will make it clearer.

```suggestion
  const [logEntriesFetchParams, setLogEntriesFetchParams] = useState(logEntriesFetchInitialParams);
```

Makes sense? Maybe I'm getting the purpose wrong and my suggestion needs to change as well

Reply:
A change in the `dependencies` actually does re-run the `useFetchEntriesEffect` hook, though depending on the how the parameters change it doesn't necessarily make another request. I'm working on refactoring a bunch of this stuff to fit in closer with existing hooks patterns, so let me see if I can get these to behave in a way that's easier to describe